lib/auth/auth_with_roles.go

```diff
@@ -536,18 +536,167 @@ func (a *ServerWithRoles) GenerateHostCerts(ctx context.Context, req *proto.Host
+			}
+			if !r.IsLocalService() {
+				return nil, trace.AccessDenied("additional system role %q cannot be applied (not a builtin service role)", r)
+			}
+		}
+
+		// load system role assertions if relevant
+		var assertions proto.UnstableSystemRoleAssertionSet
+		if req.UnstableSystemRoleAssertionID != "" {
+			assertions, err = a.authServer.UnstableGetSystemRoleAssertions(ctx, req.HostID, req.UnstableSystemRoleAssertionID)
+			if err != nil {
+				// include this error in the logs, since it might be indicative of a bug if it occurs outside of the context
+				// of a general backend outage.
+				log.Warnf("Failed to load system role assertion set %q for instance %q: %v", req.UnstableSystemRoleAssertionID, req.HostID, err)
+				return nil, trace.AccessDenied("failed to load system role assertion set with ID %q", req.UnstableSystemRoleAssertionID)
+			}
+		}
+
+		// check if additional system roles are permissible
+		for _, requestedRole := range req.SystemRoles {
+			if a.hasBuiltinRole(string(requestedRole)) {
+				// instance is already known to hold this role
+				continue
+			}
+
+			for _, assertedRole := range assertions.SystemRoles {
+				if requestedRole == assertedRole {
+					// instance recently demonstrated that it holds this role
+					continue
+				}
+			}
+
+			return nil, trace.AccessDenied("additional system role %q cannot be applied (not authorized)", requestedRole)
+		}
+
+	} else {
+		if len(req.SystemRoles) != 0 {
+			return nil, trace.AccessDenied("additional system role encoding not supported for certs of type %q", req.Role)
+		}
+	}
+
+	// prohibit privilege escalations through role changes (except the instance cert exception, handled above).
+	if !a.hasBuiltinRole(string(req.Role)) && req.Role != types.RoleInstance {
 		return nil, trace.AccessDenied("roles do not match: %v and %v", existingRoles, req.Role)
 	}
 	return a.authServer.GenerateHostCerts(ctx, req)
 }
 
+func (a *ServerWithRoles) UnstableAssertSystemRole(ctx context.Context, req proto.UnstableSystemRoleAssertion) error {
+	role, ok := a.context.Identity.(BuiltinRole)
+	if !ok || !role.IsServer() {
+		return trace.AccessDenied("system role assertions can only be executed by a teleport built-in server")
+	}
+
+	if req.ServerID != role.GetServerID() {
+		return trace.AccessDenied("system role assertions do not support impersonation (%q -> %q)", role.GetServerID(), req.ServerID)
+	}
+
+	if !a.hasBuiltinRole(string(req.SystemRole)) {
+		return trace.AccessDenied("cannot assert unheld system role %q", req.SystemRole)
+	}
+
+	if !req.SystemRole.IsLocalService() {
+		return trace.AccessDenied("cannot assert non-service system role %q", req.SystemRole)
+	}
+
+	return a.authServer.UnstableAssertSystemRole(ctx, req)
+}
+
+func (a *ServerWithRoles) RegisterInventoryControlStream(ics client.UpstreamInventoryControlStream) error {
+	// Ensure that caller is a teleport server
+	role, ok := a.context.Identity.(BuiltinRole)
+	if !ok || !role.IsServer() {
```

Comment:
Refactored but ended up reverting.  Turns out that we almost always need to perform this typecast anyway since we proceed to access fields that are unique to `BuiltinRole`, so a helper function didn't actually reduce verbosity at all, and increased the amount of typecasts that we had to perform.